src/exec/exec.go

```diff
@@ -0,0 +1,78 @@
+package exec
+
+import (
+	"context"
+	"math"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"gopkg.in/op/go-logging.v1"
+
+	"github.com/thought-machine/please/src/core"
+	"github.com/thought-machine/please/src/fs"
+	"github.com/thought-machine/please/src/process"
+)
+
+var log = logging.MustGetLogger("exec")
+
+func Exec(state *core.BuildState, label core.BuildLabel, cmdArgs []string, shareNetwork bool) {
+	exec(context.Background(), state, label, cmdArgs, shareNetwork)
+}
+
+func exec(ctx context.Context, state *core.BuildState, label core.BuildLabel, cmdArgs []string, shareNetwork bool) error {
+	target := state.Graph.TargetOrDie(label)
+
+	if !target.IsBinary && len(cmdArgs) == 0 {
+		log.Fatalf("Target %s cannot be executed; it's not marked as binary", target.Label)
+	}
+
+	var cmd string
```

Comment:
I'm always a bit dubious about this kind of `var` above an if pattern. Generally would suggest pulling it out to another function so you can use `:=` (ideally ending up with more of a single-assignment form)

Reply:
I guess it can get messy. I like the suggestion.